File: scripts/compute-nme-portfolio-opt.js
/**
 * Compute NME Portfolio Optimization
 * Markowitz Efficient Frontier for NME assets
 */
require('dotenv').config({ path: '.env.local' });
const { Client } = require('pg');

async function computeNMEPortfolioOpt() {
  const client = new Client({ connectionString: process.env.DATABASE_URL });
  await client.connect();
  console.log('Connected to database');

  // Get NME data from the view
  const nmeQuery = `
    SELECT
      nme_id,
      nme_code,
      nme_name,
      therapeutic_area,
      molecule_type,
      nme_status,
      project_count,
      task_completion_pct,
      total_ev,
      total_bac,
      portfolio_spi,
      portfolio_cpi,
      ev_ratio,
      completion_ratio
    FROM v_nme_portfolio_metrics
  `;

  const { rows: nmes } = await client.query(nmeQuery);
  console.log(`Found ${nmes.length} NMEs`);

  // Clear existing data
  await client.query('DELETE FROM nme_portfolio_opt');
  await client.query('DELETE FROM nme_frontier_curve');
  console.log('Cleared existing NME optimization data');

  // Compute return and risk scores
  const scored = nmes.map(n => {
    const portfolioSpi = parseFloat(n.portfolio_spi) || 1;
    const portfolioCpi = parseFloat(n.portfolio_cpi) || 1;
    const evRatio = parseFloat(n.ev_ratio) || 0;
    const completionRatio = parseFloat(n.completion_ratio) || 0;

    // Return = weighted combination of EV ratio and task completion
    const returnScore = 0.6 * evRatio + 0.4 * completionRatio;

    // Risk = 0.5 * (1 - SPI) + 0.5 * (1 - CPI)
    const riskScore = Math.max(0, Math.min(1, 0.5 * (1 - portfolioSpi) + 0.5 * (1 - portfolioCpi)));

    return {
      ...n,
      return_score: returnScore,
      risk_score: riskScore,
    };
  });

  // Determine efficient frontier (Pareto-optimal set)
  for (const n of scored) {
    n.is_frontier = !scored.some(other =>
      other.nme_id !== n.nme_id &&
      other.return_score >= n.return_score &&
      other.risk_score <= n.risk_score &&
      (other.return_score > n.return_score || other.risk_score < n.risk_score)
    );
  }

  // Rank frontier NMEs
  const frontierNMEs = scored.filter(n => n.is_frontier);
  frontierNMEs.sort((a, b) => a.risk_score - b.risk_score);
  frontierNMEs.forEach((n, idx) => {
    n.frontier_rank = idx + 1;
  });

  // Compute combined score and recommendation
  const medianScore = [...scored]
    .map(n => 0.5 * n.return_score + 0.5 * (1 - n.risk_score))
    .sort((a, b) => a - b)[Math.floor(scored.length / 2)] || 0.5;

  for (const n of scored) {
    n.combined_score = 0.5 * n.return_score + 0.5 * (1 - n.risk_score);

    if (n.is_frontier && n.combined_score >= medianScore * 1.2) {
      n.recommendation = 'SELECT';
    } else if (n.is_frontier) {
      n.recommendation = 'CONSIDER';
    } else if (n.combined_score >= medianScore) {
      n.recommendation = 'MONITOR';
    } else {
      n.recommendation = 'DEFER';
    }
  }

  // Insert NME recommendations
  for (const n of scored) {
    await client.query(`
      INSERT INTO nme_portfolio_opt (
        nme_id, return_score, risk_score,
        is_frontier, frontier_rank,
        recommendation, combined_score, computed_at
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, NOW())
    `, [
      n.nme_id,
      n.return_score.toFixed(6),
      n.risk_score.toFixed(6),
      n.is_frontier,
      n.frontier_rank || null,
      n.recommendation,
      n.combined_score.toFixed(6),
    ]);
  }

  console.log(`✓ Inserted ${scored.length} NME recommendations`);

  // Generate efficient frontier curve points
  // Interpolate between frontier points for smooth curve
  if (frontierNMEs.length >= 2) {
    const curvePoints = [];

    // Add actual frontier points
    for (const n of frontierNMEs) {
      curvePoints.push({
        risk: n.risk_score,
        return: n.return_score,
      });
    }

    // Interpolate additional points for smoother curve
    for (let i = 0; i < frontierNMEs.length - 1; i++) {
      const p1 = frontierNMEs[i];
      const p2 = frontierNMEs[i + 1];

      // Add 2 intermediate points
      for (let t = 0.33; t < 1; t += 0.33) {
        curvePoints.push({
          risk: p1.risk_score + t * (p2.risk_score - p1.risk_score),
          return: p1.return_score + t * (p2.return_score - p1.return_score),
        });
      }
    }

    // Sort by risk
    curvePoints.sort((a, b) => a.risk - b.risk);

    // Insert curve points
    for (const point of curvePoints) {
      await client.query(`
        INSERT INTO nme_frontier_curve (risk_value, return_value, computed_at)
        VALUES ($1, $2, NOW())
      `, [point.risk.toFixed(6), point.return.toFixed(6)]);
    }

    console.log(`✓ Inserted ${curvePoints.length} frontier curve points`);
  }

  const tierCounts = scored.reduce((acc, n) => {
    acc[n.recommendation] = (acc[n.recommendation] || 0) + 1;
    return acc;
  }, {});

  console.log('Tier distribution:', tierCounts);
  console.log(`Frontier NMEs: ${frontierNMEs.length}`);

  await client.end();
  console.log('Done');
}

computeNMEPortfolioOpt().catch(console.error);

File: scripts/compute_nme_portfolio_opt.py
"""
NME Portfolio — Markowitz Efficient Frontier Optimisation
=========================================================
Treats each active NME compound as a portfolio asset and applies
classical Markowitz mean-variance optimisation to identify the
optimal allocation across the compound pipeline.

Framework:
  1. Return score per NME  = composite of EV ratio + task completion + phase maturity
  2. Risk score per NME    = composite of EVM risk + phase risk + status risk
  3. Covariance matrix     = therapeutic-area-driven correlation structure
  4. Monte Carlo (50,000)  = random portfolios plotted on risk/return space
  5. scipy frontier curve  = exact min-variance at 60 target return levels
  6. Max-Sharpe portfolio  = optimal allocation weights
  7. Min-variance portfolio = most defensive allocation weights

Output tables:
  nme_portfolio_opt     — per-NME tier / weights / Sharpe ratio
  nme_frontier_curve    — (risk, return) points tracing the frontier

Run:
  DATABASE_URL=... python3.11 scripts/compute_nme_portfolio_opt.py
"""

import os
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import psycopg2
import psycopg2.extras
from scipy.optimize import minimize

# ── Database ───────────────────────────────────────────────────────────────────
DB_URL = os.environ.get(
    "DATABASE_URL",
    "postgresql://claudegabriel@localhost:5433/nme_portfolio"
)

def get_conn():
    return psycopg2.connect(DB_URL)

NME_METRICS_SQL = """
SELECT
    nme_id, nme_name, nme_code,
    therapeutic_area, molecule_type, nme_status,
    total_bac,
    ev_ratio, completion_ratio,
    phase_risk, status_risk, evm_risk
FROM v_nme_portfolio_metrics
"""

# ── 1. Composite return / risk scores ─────────────────────────────────────────
RISK_FREE_RATE = 0.05   # notional baseline for Sharpe calculation

def compute_scores(df: pd.DataFrame) -> pd.DataFrame:
    df = df.copy()
    for col in ["ev_ratio", "completion_ratio", "phase_risk",
                "status_risk", "evm_risk"]:
        df[col] = df[col].apply(lambda x: float(x) if x is not None else 0.0)

    # return_score: how much value is being delivered vs budgeted
    df["return_score"] = (
        0.50 * df["ev_ratio"].clip(0, 1) +
        0.30 * df["completion_ratio"].clip(0, 1) +
        0.20 * (1 - df["phase_risk"].clip(0, 1))   # phase maturity → higher return
    ).round(4)

    # risk_score: weighted blend of execution, phase, and regulatory risk
    df["risk_score"] = (
        0.50 * df["evm_risk"].clip(0, 1) +
        0.30 * df["phase_risk"].clip(0, 1) +
        0.20 * df["status_risk"].clip(0, 1)
    ).round(4)

    df["sharpe_ratio"] = (
        (df["return_score"] - RISK_FREE_RATE) /
        df["risk_score"].clip(lower=0.01)
    ).round(4)

    return df

# ── 2. Covariance matrix ───────────────────────────────────────────────────────
def build_covariance(df: pd.DataFrame) -> np.ndarray:
    """
    Σ_ii  = risk_i²
    Σ_ij  = ρ_ij × risk_i × risk_j
    ρ_ij  = 0.55 if same therapeutic area, else 0.15
    """
    n = len(df)
    risks = df["risk_score"].values
    tas   = df["therapeutic_area"].values

    rho = np.full((n, n), 0.15)
    for i in range(n):
        for j in range(n):
            if i == j:
                rho[i, j] = 1.0
            elif tas[i] == tas[j]:
                rho[i, j] = 0.55

    cov = rho * np.outer(risks, risks)
    # Ensure positive-definite (add small jitter to diagonal)
    cov += np.eye(n) * 1e-8
    return cov

# ── 3. Portfolio stats helper ─────────────────────────────────────────────────
def port_stats(weights: np.ndarray, mu: np.ndarray, cov: np.ndarray) -> tuple:
    ret  = float(weights @ mu)
    var  = float(weights @ cov @ weights)
    risk = float(np.sqrt(max(var, 0)))
    sharpe = (ret - RISK_FREE_RATE) / max(risk, 1e-6)
    return ret, risk, sharpe

# ── 4. Monte Carlo simulation ─────────────────────────────────────────────────
N_SIMS = 50_000

def monte_carlo(mu: np.ndarray, cov: np.ndarray) -> pd.DataFrame:
    n = len(mu)
    results = []
    rng = np.random.default_rng(42)

    # Also include single-asset and equal-weight portfolios
    baselines = [np.eye(n)[i] for i in range(n)]
    baselines.append(np.ones(n) / n)

    for w in baselines:
        ret, risk, sh = port_stats(w, mu, cov)
        results.append({"ret": ret, "risk": risk, "sharpe": sh,
                        "weights": w.tolist()})

    # Dirichlet random weights for uniform coverage over the simplex
    alpha = np.ones(n)
    for _ in range(N_SIMS):
        w = rng.dirichlet(alpha)
        ret, risk, sh = port_stats(w, mu, cov)
        results.append({"ret": ret, "risk": risk, "sharpe": sh,
                        "weights": w.tolist()})

    return pd.DataFrame(results)

# ── 5. scipy exact efficient frontier ─────────────────────────────────────────
def compute_frontier_curve(
    mu: np.ndarray, cov: np.ndarray, n_points: int = 60
) -> list[dict]:
    n = len(mu)
    constraints = [{"type": "eq", "fun": lambda w: np.sum(w) - 1}]
    bounds = [(0, 1)] * n
    w0 = np.ones(n) / n

    target_returns = np.linspace(mu.min(), mu.max(), n_points)
    curve = []

    for target_ret in target_returns:
        con = constraints + [{"type": "eq", "fun": lambda w, r=target_ret: w @ mu - r}]
        res = minimize(
            lambda w: float(w @ cov @ w),   # minimise variance
            w0,
            method="SLSQP",
            bounds=bounds,
            constraints=con,
            options={"ftol": 1e-12, "maxiter": 1000},
        )
        if res.success:
            w_opt = np.array(res.x).clip(0, 1)
            w_opt /= w_opt.sum()
            _, risk, _ = port_stats(w_opt, mu, cov)
            curve.append({
                "portfolio_risk": round(float(risk), 6),
                "portfolio_return": round(float(target_ret), 6),
                "is_min_variance": False,
                "is_max_sharpe": False,
            })

    return curve

# ── 6. Identify special portfolios ────────────────────────────────────────────
def find_optimal_portfolios(
    mc_df: pd.DataFrame,
) -> tuple[int, int]:
    """Returns indices of max-Sharpe and min-variance portfolios."""
    idx_sharpe  = mc_df["sharpe"].idxmax()
    idx_min_var = mc_df["risk"].idxmin()
    return int(idx_sharpe), int(idx_min_var)

# ── 7. Tier assignment ────────────────────────────────────────────────────────
def assign_tier(sharpe: float, p75: float, p50: float, p25: float) -> str:
    """
    Sharpe-ratio percentile tiers (more informative than portfolio-weight tiers
    when one asset dominates the max-Sharpe portfolio):
      CORE      = top 25% Sharpe  (best risk-adjusted returns)
      GROWTH    = 25–50% Sharpe   (above-median risk-adjusted returns)
      SATELLITE = 50–75% Sharpe   (below-median but positive contribution)
      EXCLUDE   = bottom 25%      (risk-adjusted underperformers)
    """
    if sharpe >= p75:
        return "CORE"
    if sharpe >= p50:
        return "GROWTH"
    if sharpe >= p25:
        return "SATELLITE"
    return "EXCLUDE"

# ── 8. Main ───────────────────────────────────────────────────────────────────
def main():
    print("Connecting to database…")
    conn = get_conn()
    cur  = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)

    cur.execute(NME_METRICS_SQL)
    rows = cur.fetchall()
    df = pd.DataFrame(rows)
    print(f"  Loaded {len(df)} NMEs from v_nme_portfolio_metrics")

    if df.empty:
        print("No NMEs found."); conn.close(); return

    df = compute_scores(df)
    mu  = df["return_score"].values.astype(float)
    cov = build_covariance(df)

    print(f"\n  Return scores: min={mu.min():.3f}  max={mu.max():.3f}  "
          f"mean={mu.mean():.3f}")
    print(f"  Risk scores:   min={df['risk_score'].min():.3f}  "
          f"max={df['risk_score'].max():.3f}  mean={df['risk_score'].mean():.3f}")

    # ── Monte Carlo ──────────────────────────────────────────────────────────
    print(f"\nRunning {N_SIMS:,} Monte Carlo simulations…")
    mc_df = monte_carlo(mu, cov)
    print(f"  Done. Best Sharpe in simulation: {mc_df['sharpe'].max():.4f}")

    idx_sharpe, idx_min_var = find_optimal_portfolios(mc_df)
    opt_weights    = np.array(mc_df.loc[idx_sharpe,  "weights"])
    min_var_weights= np.array(mc_df.loc[idx_min_var, "weights"])

    opt_ret,  opt_risk,  opt_sh  = port_stats(opt_weights,     mu, cov)
    mvp_ret,  mvp_risk,  mvp_sh  = port_stats(min_var_weights, mu, cov)

    print(f"\n  Max-Sharpe portfolio:   return={opt_ret:.3f}  "
          f"risk={opt_risk:.3f}  sharpe={opt_sh:.4f}")
    print(f"  Min-Variance portfolio: return={mvp_ret:.3f}  "
          f"risk={mvp_risk:.3f}  sharpe={mvp_sh:.4f}")

    # ── Exact frontier curve ─────────────────────────────────────────────────
    print("\nComputing exact efficient frontier (scipy SLSQP)…")
    frontier_pts = compute_frontier_curve(mu, cov, n_points=60)
    print(f"  Solved {len(frontier_pts)} frontier points")

    # Mark special points on curve
    if frontier_pts:
        min_var_pt = min(frontier_pts, key=lambda p: p["portfolio_risk"])
        min_var_pt["is_min_variance"] = True
        # Max-Sharpe on frontier: highest (return - Rf) / risk
        max_sh_pt = max(
            frontier_pts,
            key=lambda p: (p["portfolio_return"] - RISK_FREE_RATE) /
                          max(p["portfolio_risk"], 1e-6)
        )
        max_sh_pt["is_max_sharpe"] = True

    # ── Assign NME-level metrics ─────────────────────────────────────────────
    df["optimal_weight"]  = opt_weights.round(6)
    df["min_var_weight"]  = min_var_weights.round(6)
    df["is_frontier"]     = df["optimal_weight"] > 0.005

    p25 = float(df["sharpe_ratio"].quantile(0.25))
    p50 = float(df["sharpe_ratio"].quantile(0.50))
    p75 = float(df["sharpe_ratio"].quantile(0.75))
    df["tier"] = df["sharpe_ratio"].apply(
        lambda s: assign_tier(s, p75, p50, p25)
    )

    # ── Summary ──────────────────────────────────────────────────────────────
    tier_counts = df["tier"].value_counts()
    print(f"\n  Tier distribution — "
          f"CORE:{tier_counts.get('CORE',0)}  "
          f"GROWTH:{tier_counts.get('GROWTH',0)}  "
          f"SATELLITE:{tier_counts.get('SATELLITE',0)}  "
          f"EXCLUDE:{tier_counts.get('EXCLUDE',0)}")

    # ── Upsert NME opt results ───────────────────────────────────────────────
    print("\nWriting nme_portfolio_opt…")
    upsert_sql = """
    INSERT INTO nme_portfolio_opt
        (nme_id, return_score, risk_score, sharpe_ratio,
         is_frontier, optimal_weight, min_var_weight, tier,
         ev_ratio, evm_risk, phase_risk, computed_at)
    VALUES
        (%(nme_id)s, %(return_score)s, %(risk_score)s, %(sharpe_ratio)s,
         %(is_frontier)s, %(optimal_weight)s, %(min_var_weight)s, %(tier)s,
         %(ev_ratio)s, %(evm_risk)s, %(phase_risk)s, NOW())
    ON CONFLICT (nme_id) DO UPDATE SET
        return_score    = EXCLUDED.return_score,
        risk_score      = EXCLUDED.risk_score,
        sharpe_ratio    = EXCLUDED.sharpe_ratio,
        is_frontier     = EXCLUDED.is_frontier,
        optimal_weight  = EXCLUDED.optimal_weight,
        min_var_weight  = EXCLUDED.min_var_weight,
        tier            = EXCLUDED.tier,
        ev_ratio        = EXCLUDED.ev_ratio,
        evm_risk        = EXCLUDED.evm_risk,
        phase_risk      = EXCLUDED.phase_risk,
        computed_at     = EXCLUDED.computed_at
    """
    records = df[[
        "nme_id", "return_score", "risk_score", "sharpe_ratio",
        "is_frontier", "optimal_weight", "min_var_weight", "tier",
        "ev_ratio", "evm_risk", "phase_risk",
    ]].to_dict("records")

    for rec in records:
        rec["is_frontier"] = bool(rec["is_frontier"])
        for k in ["return_score","risk_score","sharpe_ratio","optimal_weight",
                  "min_var_weight","ev_ratio","evm_risk","phase_risk"]:
            rec[k] = float(rec[k])

    cur2 = conn.cursor()
    cur2.executemany(upsert_sql, records)

    # ── Write frontier curve ─────────────────────────────────────────────────
    print("Writing nme_frontier_curve…")
    cur2.execute("DELETE FROM nme_frontier_curve")   # truncate old curve
    if frontier_pts:
        cur2.executemany("""
            INSERT INTO nme_frontier_curve
                (portfolio_risk, portfolio_return, is_min_variance, is_max_sharpe, computed_at)
            VALUES
                (%(portfolio_risk)s, %(portfolio_return)s, %(is_min_variance)s,
                 %(is_max_sharpe)s, NOW())
        """, frontier_pts)

    conn.commit()
    print(f"  Wrote {len(records)} NME rows + {len(frontier_pts)} frontier curve points")

    # ── Print ranking table ──────────────────────────────────────────────────
    top = df.sort_values("sharpe_ratio", ascending=False)
    print("\n── NME Rankings ─────────────────────────────────────────────────────────")
    print(f"{'NME':<14} {'TA':<20} {'Ret':>5} {'Risk':>5} {'Sharpe':>7} "
          f"{'OptWt':>6} {'MVWt':>6} {'Tier':<10}")
    print("─" * 80)
    for _, r in top.iterrows():
        print(f"{r['nme_code']:<14} {r['therapeutic_area']:<20} "
              f"{r['return_score']:>5.3f} {r['risk_score']:>5.3f} "
              f"{r['sharpe_ratio']:>7.4f} {r['optimal_weight']:>6.3f} "
              f"{r['min_var_weight']:>6.3f} {r['tier']:<10}")

    conn.close()
    print("\nDone.")


if __name__ == "__main__":
    np.random.seed(42)
    main()
